### PR TITLE
histogram support for ruby client

### PR DESCRIPTION
**Pull request by rolandhawk**
Is this acceptable?

ref #21

### CODE REVIEW

File: spec/prometheus/client/formats/text_spec.rb [hunk 1]
```diff
@@ -67,6 +80,13 @@
 qux{for="sake",code="1",quantile="0.99"} 15.3
 qux_sum{for="sake",code="1"} 1243.21
 qux_count{for="sake",code="1"} 93
+# TYPE xuq histogram
+# HELP xuq xuq description
+xuq{code="ah",le="10"} 1
+xuq{code="ah",le="20"} 2
+xuq{code="ah",le="30"} 2
```
Comment (brian-brazil) [inline spec/prometheus/client/formats/text_spec.rb]: You must expose a +Inf bucket wit the same value as _count

File: lib/prometheus/client/histogram.rb [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+
+          each do |bucket, count|
+            self[bucket] += 1 if value <= bucket
+          end
+        end
+      end
+
+      # Offer a way to manually specify buckets
+      def initialize(name, docstring, base_labels = {}, buckets = [])
+        @buckets = buckets
```
Comment (brian-brazil) [inline lib/prometheus/client/histogram.rb]: What are the default buckets?

File: lib/prometheus/client/histogram.rb [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+          @total += 1
+
+          each do |bucket, count|
+            self[bucket] += 1 if value <= bucket
+          end
+        end
+      end
+
+      # Offer a way to manually specify buckets
+      def initialize(name, docstring, base_labels = {}, buckets = [])
```
Comment (brian-brazil) [inline lib/prometheus/client/histogram.rb]: You need to check the buckets are ordered.

Comment (rolandhawk) [inline lib/prometheus/client/histogram.rb]: Do I need to validate buckets order and raise if not ordered 
or
Just order the buckets?

Comment (brian-brazil) [inline lib/prometheus/client/histogram.rb]: We've raised elsewhere, as it's likely a typo on the part of the user.

File: lib/prometheus/client/histogram.rb [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+      def initialize(name, docstring, base_labels = {}, buckets = DEFAULT_BUCKETS)
+        @buckets = buckets.sort
+        super(name, docstring, base_labels)
+      end
+
+      def type
+        :histogram
+      end
+
+      def add(labels, value)
```
Comment (brian-brazil) [inline lib/prometheus/client/histogram.rb]: This should be called `observe`

File: lib/prometheus/client/registry.rb [hunk 5]
```diff
@@ -43,6 +44,10 @@ def gauge(name, docstring, base_labels = {})
         register(Gauge.new(name, docstring, base_labels))
       end
 
+      def histogram(*args)
```
Comment (grobie) [inline lib/prometheus/client/registry.rb]: For documentation purposes, I'd vote to explicitly list the arguments.

File: lib/prometheus/client/histogram.rb [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+      end
+
+      # DEFAULT_BUCKETS are the default Histogram buckets. The default buckets are
+      # tailored to broadly measure the response time (in seconds) of a
+      # network service. (From DefBuckets client_golang)
+      DEFAULT_BUCKETS = [0.005, 0.01, 0.025, 0.05, 0.1, 0.25, 0.5, 1, 2.5, 5, 10]
+
+      # Offer a way to manually specify buckets
+      def initialize(name, docstring, base_labels = {}, buckets = DEFAULT_BUCKETS)
+        @buckets = buckets.sort
```
Comment (grobie) [inline lib/prometheus/client/histogram.rb]: I agree with @brian-brazil and would vote to validate and raise an error on unsorted buckets. It's easy to make a typo when specifying small buckets values. We should help the user to detect such issues.

**Comment by grobie** [comment]
Thanks for your contribution! This looks great in general, please only fix the last few comments.

I updated the formatting to the latest rubocop rules (not sure following rubocop was a good idea, they change their mind so often ...). Please rebase and squash your changes against current master. I'll fix the test dependencies for "old" ruby versions tomorrow.

**Comment by rolandhawk** [comment]
Hi @grobie @brian-brazil, I have updated my PR, please kindly check it

**Comment by grobie** [comment]
Thanks a lot @rolandhawk!

**Pull request merged**